### ISSUE TITLE
S3 Support does not work for private bucket

### ISSUE THREAD

**Issue by timonbimon** [tensorflow/tensorflow#15662]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.4.0
- **Python version**: 3.5.2
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: -
- **GPU model and memory**: - 
- **Exact command to reproduce**: -

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

When trying to access a file on one of my private S3 buckets, I get an error message that the object does not exist (see below for source code and traceback). Using the AWS CLI downloading the object works fine(aka I'm sure I have the access rights to access the bucket). Also accessing the object in a public S3 bucket (like the one in issue #15159) works fine. I tried looking into the source code [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/s3/s3_file_system.cc#L404) to see if it's maybe an issue with missing or wrong environment variables concerning the AWS credentials, but I couldn't find any code checking for any credentials at all. Am I missing something simple?

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

When running the following code

```python
from tensorflow.python.lib.io import file_io
file_io.stat('s3://myprivatebucket/filethatexists')
```

I get the following error

```
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/lib/io/file_io.py", line 98, in size
    return stat(self.__name).length
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/lib/io/file_io.py", line 540, in stat
    return file_statistics
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/errors_impl.py", line 473, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.NotFoundError: Object s3://myprivatebucket/filethatexists does not exist
```

**Comment by quaeler** [comment]
Could you try building locally @yongtang's [branch which contains more logging](https://github.com/yongtang/tensorflow/tree/15159-s3-logging) to see if there's anything more useful logged?

**Comment by timonbimon** [comment]
Hmm, I tried doing that but unfortunately don't get any more detailed logging. Is there a command I could try that should trigger a more detailed error message with @yongtang's AWS logging - just to double-check that at least that is working correctly?

**Comment by quaeler** [comment]
@yongtang should confirm, but it looks like it is obeying the standard environment variable `TF_CPP_MIN_LOG_LEVEL` - if you have this unset or set to 0 it should dump the most amount of logging. If that's already your case, maybe you can add a logging in `AWSLogSystem::InitializeAWSLogging` in `tensorflow/core/platform/s3/aws_logging.cc` to note that it has been spun up.

**Comment by yongtang** [comment]
@timonbimon The following is the expected output with AWS logging enabled (with all default value without customerization). Please see if you could get something similar.
```
ubuntu@ubuntu:~$ python -c "from tensorflow.python.lib.io import file_io; file_io.stat('s3://datasets.elasticmapreduce/ngrams/books/20090715/eng-1M/1gram/data')"
2017-12-28 09:49:22.726196: I tensorflow/core/platform/s3/aws_logging.cc:53] Initializing Curl library
2017-12-28 09:49:23.134517: I tensorflow/core/platform/s3/aws_logging.cc:53] Initializing config loader against fileName /home/ubuntu//.aws/config and using profilePrefix = 1
2017-12-28 09:49:23.134559: I tensorflow/core/platform/s3/aws_logging.cc:53] Initializing config loader against fileName /home/ubuntu//.aws/credentials and using profilePrefix = 0
2017-12-28 09:49:23.134570: I tensorflow/core/platform/s3/aws_logging.cc:53] Setting provider to read credentials from /home/ubuntu//.aws/credentials for credentials file and /home/ubuntu//.aws/config for the config file , for use with profile default
2017-12-28 09:49:23.134579: I tensorflow/core/platform/s3/aws_logging.cc:53] Creating HttpClient with max connections2 and scheme http
2017-12-28 09:49:23.134596: I tensorflow/core/platform/s3/aws_logging.cc:53] Initializing CurlHandleContainer with size 2
2017-12-28 09:49:23.134607: I tensorflow/core/platform/s3/aws_logging.cc:53] Creating Instance with default EC2MetadataClient and refresh rate 900000
2017-12-28 09:49:23.134653: I tensorflow/core/platform/s3/aws_logging.cc:53] Successfully reloaded configuration.
2017-12-28 09:49:23.134702: I tensorflow/core/platform/s3/aws_logging.cc:53] Initializing CurlHandleContainer with size 25
2017-12-28 09:49:23.134896: I tensorflow/core/platform/s3/aws_logging.cc:53] Pool grown by 2
2017-12-28 09:49:23.134914: I tensorflow/core/platform/s3/aws_logging.cc:53] Connection has been released. Continuing.
2017-12-28 09:49:23.419656: I tensorflow/core/platform/s3/aws_logging.cc:53] Connection has been released. Continuing.
2017-12-28 09:49:23.491859: I tensorflow/core/platform/s3/aws_logging.cc:53] Cleaning up CurlHandleContainer.
2017-12-28 09:49:23.491891: I tensorflow/core/platform/s3/aws_logging.cc:53] Cleaning up CurlHandleContainer.
ubuntu@ubuntu:~$ 
```

**Comment by timonbimon** [comment]
ah that was my bad. I git-cloned @yongtang's repository and then forgot to switch the branch.
The logging works (great stuff @yongtang 👍) and the error was 
`The authorization header is malformed; the region 'us-east-1' is wrong; expecting 'eu-west-1'`

The fix that worked for me was to explicitly do an `export AWS_REGION=eu-west-1`.

However the region set in my ~/.aws/config was already 'eu-west-1' so there seems to be an error that this value wasn't taken by default (or it was overwritten later on when the request was made).

**Comment by asimshankar** [comment]
Thanks for digging into this @quaeler @yongtang, glad @timonbimon could find a solution.

@jhseu FYI

From a cursory reading of https://github.com/tensorflow/tensorflow/blob/r1.5/tensorflow/core/platform/s3/s3_file_system.cc#L40 - it seems that only the environment variables are used and the configuration in the filesystem is not accessed. But I could be wrong.

Contributions to address this are welcome!

**Comment by yongtang** [comment]
The AWS C++ SDK does not use the config file by default though it is possible to load the config file explicitly. Added a PR #15723 for the fix. Please take a look if interested.

**Comment by jvishnuvardhan** [comment]
Automatically closing this out since I understand it to be resolved by the PR #15723 (merged already), but please let me know if I'm mistaken.Thanks!
